packages/nimble-components/src/table/models/table-validator.spec.ts

```diff
@@ -1,234 +1,316 @@
+import type { TableRecord, TableValidity } from '../types';
 import { TableValidator } from './table-validator';
 
 describe('TableValidator', () => {
-    it('setting valid field for ID is valid', () => {
-        const data = [
-            { stringField: 'value-1', numberField: 10 },
-            { stringField: 'value-2', numberField: 11 }
-        ];
-        const validator = new TableValidator();
+    let validator: TableValidator<TableRecord>;
 
-        const isValid = validator.validateRecordIds(data, 'stringField');
-        expect(isValid).toBeTrue();
-        expect(validator.isValid()).toBeTrue();
-        const validity = validator.getValidity();
-        expect(validity.duplicateRecordId).toBeFalse();
-        expect(validity.invalidRecordId).toBeFalse();
-        expect(validity.missingRecordId).toBeFalse();
+    beforeEach(() => {
+        validator = new TableValidator();
     });
 
-    it('setting `undefined` field for ID is valid', () => {
-        const data = [
-            { stringField: 'value-1', numberField: 10 },
-            { stringField: 'value-2', numberField: 11 }
-        ];
-        const validator = new TableValidator();
+    function validateValidity(invalidKeys: (keyof TableValidity)[]): void {
```

Comment:
I have a few concerns with this pattern:
- This helper function is coupled to shared state unnecessarily (could be more standalone with validity as a param)
- We should try and keep `expect` clauses in `it` tests as much as possible. So you can see the assertions a test is making in the test itself. Helper functions should return values that you can `expect` on.
- The `expect` cases are all very conditional / dynamic, it makes it hard to understand the expected result.

I think we could have a helper like `validateValidity(validity): (keyof TableValidity)[]` that given a validity returns an array of keys of properties that are failing and in the test do:
 ```
expect(validateValidity(validity)).toEqual(jasmine.arrayWithExactContents(['duplicateRecordId']));
expect(validateValidity(validity)).to
```

arrayWithExactContents compares contents, not order: https://jasmine.github.io/api/edge/jasmine.html#.arrayWithExactContents